src/firefly/js/visualize/projection/WavelengthHeaderParser.js

```diff
@@ -19,21 +21,147 @@ const allGoodValues= (...numbers)  => numbers.every((n) => !isNaN(n));
 
 const L_10= Math.log(10);
 
-function calculateWavelengthParams(parse, altWcs, wlTable) {
+/**
+ * According to A&A 395, 1061-1075 (2002) DOI: 10.1051/0004-6361:20021326
+ * Representations of world coordinates in FITS E. W. Greisen1 - M. R. Calabretta2
+ * paper,  the CD_ij is for the older FITs header. The newer FITs header has PC only.
+ * 1. If both PC and CD exist, it is wrong.  Exception is thrown
+ * 2. If PC is not exist, one or more CD is exist, use CD_ij (j=1..N); if any of CD_ij
+ *   is not defined, the default is 0.0;
+ * 3. If any PC_ij is defined or neigther PC nor CD is defined, we use PC, the default
+ *    is 0 for j!=i and 1 if j==i
+ *
+ * @param header
+ * @param zeroHeader
+ * @param which
+ * @returns {string}
+ */
+function getPC_ijKey(header, zeroHeader,which){
+
+    var key;
+    let hasPC=false, hasCD=false;
+    for (key in header) {
+        if (key.startsWith(`PC${which}_`)) {
+            hasPC = true;
+            break;
+        }
+    }
+    if (!hasPC && zeroHeader){
+        for (key in zeroHeader) {
+            if (key.startsWith(`PC${which}_`)) {
+                hasPC = true;
+                break;
+            }
+        }
+    }
+    for (key in header) {
+        if (key.startsWith(`CD${which}_`)) {
+            hasCD = true;
+            break;
+        }
+    }
+    if (!hasCD && zeroHeader){
+        for (key in zeroHeader) {
+            if (key.startsWith(`CD${which}_`)) {
+                hasCD = true;
+                break;
+            }
+        }
+    }
+    if (hasPC && hasCD){
+        throw Error('The FITs header contains both PC_ij and CD_ij');
+    }
+    if (!hasPC && hasCD){
+        return 'CD';
+    }
+    return 'PC';
+}
+
+
+/**
+ * If the FITs header has 'WCSAXES', this will be the wavelength axis.
+ * If 'WCSAXES' is not defined, the default will be the larger of naxis and the j where j=1, 2, 3, ..)
+ * @param parse
+ * @returns {*}
+ */
+function getWCSAXES(parse){
+    const  wcsAxis = parse.getValue('WCSAXES');
+
+    if (wcsAxis) return wcsAxis;
+
+    const nAxis= parse.getIntValue('NAXIS');
+
+    const ctype= parse.getValue(`CTYPE${nAxis+1}`, undefined);
+
+    if(ctype){
+        return (nAxis+1).toString();
+    }
+    else{
+        return nAxis.toString();
+    }
+
+}
+function assignDefault(inArr, keyRoot, which, N){
```

Comment:
i would be better to name this routine more clearly.  It look like a general reusable name to it has a specific purpose.

Also you might want to document it. I am having trouble understanding what it is doing.